### DISCUSSION TITLE
Additional solver output

### DISCUSSION THREAD

**Discussion by DKenefake** [qpsolvers/qpsolvers#106] [Ideas]
Howdy, 

For a lot of my work, I usually need additional aspects of the solution in addition to the optimal point, such as the active set, and dual variables.

I was wondering if there was any appetite to add something of the following nature as an optional return from the solve function? Such as a keyword arg to export a SolverOutput object instead of the optimal point? I have implementations for gurobi and quadprog in a separate project.

```python
@dataclass
class SolverOutput:
    """
    This is the generic Solver information object. This will be the general return object from all of the back end \\
    solvers. This was done to remove the need for the user to specialize IO for any particular Solver. It contains\\
    all of the information you would need for the optimization solution including, optimal value, optimal solution, the \\
    active set, the value of the slack variables, and the Lagrange multipliers associated with every constraint (these are listed)\\
     as the dual variables.
    Members:
    obj: objective value of the optimal solution \n
    sol: x*, numpy array \n
    Optional Parameters -> None or numpy.ndarray type
    slack: the slacks associated with every constraint \n
    equality_indices: the active set of the solution, including strongly and weakly active constraints \n
    dual: the lagrange multipliers associated with the problem\n
    """
    obj: float
    sol: numpy.ndarray

    slack: Optional[numpy.ndarray]
    active_set: Optional[numpy.ndarray]
    dual: Optional[numpy.ndarray]
```

**Comment by stephane-caron** [comment]
Absolutely :+1:  There is an appetite for this feature, it was e.g. asked in https://github.com/stephane-caron/qpsolvers/discussions/35 

It makes full sense to me. I expect most of the work to be in piping out the proper outputs for each solver, then typecheck and document it (e.g. some solvers may not fill out all the extra fields).

If you're in let's start a PR :smiley:

**Comment by DKenefake** [comment]
Sounds good! I can start in earnest next week on this.

**Comment by stephane-caron** [comment]
I can help with e.g. the typing or unit testing to integrate with the rest of the lib, feel free to send some ``TODO(scaron): ...`` my way :smiley:

**Comment by stephane-caron** [comment]
> I have implementations for gurobi and quadprog in a separate project.

@DKenefake Are these projects open source, and if so can you share pointers to the code?

I may have some time in the upcoming month to take a look and help move the PR forward.

**Comment by stephane-caron** [comment]
We can include this feature in the next major version of the project. Here is a note to discuss the API: https://github.com/stephane-caron/qpsolvers/discussions/73

> I have implementations for gurobi and quadprog in a separate project.

@DKenefake If you don't have time to iterate on https://github.com/stephane-caron/qpsolvers/pull/55 (which I can perfectly understand :ok_hand:), would you mind sharing your code for gurobi and quadprog?

I may be able to work on this in the upcoming months.

**Comment by stephane-caron** [comment]
And finally the feature has made it to [v2.7.0](https://github.com/stephane-caron/qpsolvers/releases/tag/v2.7.0), including all solvers to the party :smiley:
